components/causal_ts/src/lib.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+#[macro_use]
+extern crate tikv_util;
+
+mod errors;
+pub use errors::*;
+mod tso;
+pub use tso::*;
+mod metrics;
+pub use metrics::*;
+mod observer;
+
+pub use observer::*;
+
+use crate::errors::Result;
+use txn_types::TimeStamp;
+
+/// Trait of causal timestamp provider.
+pub trait CausalTsProvider: Send + Sync {
+    /// Get a new timestamp.
+    fn get_ts(&self) -> Result<TimeStamp>;
```

Comment:
Perhaps it can accept an additional API to indicate that next returned TS should not be less than given value. So the existing batch doesn't have to be abandoned.

Reply:
You mean the `flush()` method ?
The only scenario invokes `flush()` now is leader transfer. But we do not know the TS of the original leader.

Reply:
The approach I think of is after transferring leader, the region will get a latest tso without clearing current batch, and then will use that tso as parameter to require tso for requests. So that regions without leader change can still using the existing batches, and with leader changes will use the fresher tso.

It's an improvement and not required in this PR.

Reply:
Got it ~
Then can we implement [batching up](https://github.com/tikv/tikv/pull/12099#discussion_r832135245) with this improvement in another PR ?